pueblo/ngr/runner.py

```diff
@@ -455,22 +453,66 @@ def test(self) -> None:
         """
         Test a Python thing, based on which test runner is installed.
 
-        TODO: Figure out how to invoke target `poe check`, which is popular amongst
-              users of `pueblo`. It bundles linter/checkstyle and software test targets
-              into a single entrypoint, similar to what `gradle check` is doing.
+        First, try a high-level target of some sort of management tool,
+        discover its targets, and invoke `check` or `test`.
+
+        When no high-level tool can be discovered, just invoke `pytest`.
         """
 
+        # On Poetry projects, invoke poetry/pytest.
+        # TODO: Also allow using Poetry with other task runners, see below.
         if self.has_poetry_lock:
             # TODO: poetry run which pytest
             run_command("poetry run pytest --config-file=.")
 
         else:
+            uses_poe = self.has_pyproject_toml and self.pyproject_has_string("[tool.poe.tasks]")
             has_pytest = shutil.which("pytest") is not None
-            if has_pytest:
+
+            # When a `pyproject.toml` file includes `poethepoet` task definitions, try to
+            # invoke sensible targets like `check` or `test`. This is popular amongst users
+            # of `pueblo` and friends.
+            # `poe check` bundles linter/checkstyle and software test targets into a single
+            # entrypoint, similar to how `gradle check` is doing it.
+            if uses_poe:
+                candidates = ["check", "test"]
+                poe_tasks = self.get_poe_tasks()
+                success = False
+                for candidate in candidates:
+                    if candidate in poe_tasks:
+                        run_command(f"poe {candidate}")
+                        success = True
+                        break
+                if not success:
+                    raise RuntimeError(f"Failed to discover poe task from candidates: {candidates}")
+
+            elif has_pytest:
                 run_command("pytest")
             else:
                 raise NotImplementedError("No handler to invoke Python item")
 
+    def pyproject_has_string(self, needle: str) -> bool:
+        """
+        Check whether `pyproject.toml` file contains given string.
+        """
+        pyproject_toml_file = self.path / "pyproject.toml"
+        pyproject_toml = pyproject_toml_file.read_text()
+        return needle in pyproject_toml
+
+    def get_poe_tasks(self) -> t.List[str]:
+        """
+        From `poethepoet._list_tasks`.
+        """
+        try:
+            from poethepoet.config import PoeConfig
+
+            config = PoeConfig()
+            config.load()
+            return [task for task in config.tasks.keys() if task and task[0] != "_"]
+        except Exception as ex:  # pylint: disable=broad-except
+            # this happens if there's no pyproject.toml present
+            raise ValueError(f"Discovering poe task names failed: {ex}") from ex
```

Comment:
Dear @nat-n,

I haven't been able to use `poethepoet._list_tasks()` 1:1, because it uses the `print()` function. Would you consider providing such a convenience function via `poethepoet.list_tasks()`, so that external tools like `ngr` can easily inquire the list of tasks as a `t.List[str]`?

With kind regards,
Andreas.

Reply:
Hi @amotl,
Sure, what do you think about refactoring the `_list_tasks` method to factor our a new method in the same module like the following which you could also use?

```
def iter_tasks(project_root: str = "."):
    from .config import PoeConfig

    config = PoeConfig(cwd=project_root)
    config.load()
    for task in config.tasks.keys():
        if task and task[0] != "_":
            yield task
```

Feel free to open a PR. 

---
That said, you could also get the same benefit with less dependency with along the lines of:

```
def load_poe_tasks(cwd: str = "."):
    try:
        import tomllib as tomli
    except ImportError:
        import tomli  # type: ignore[no-redef]
    
    with path.open("rb") as file:
        return list(tomli.load(file).get("tool",{}).get("poe",{}).get("tasks",{}).keys())
```

Up to you what you prefer.

Also of potential interest:
- How poe manages running a subprocess with a virtualenv: https://github.com/nat-n/poethepoet/blob/43cc75b20c96002d9b4e920ff1a30c8d6b755a1f/poethepoet/virtualenv.py#L8
- How poe looks for a pyproject.toml file: https://github.com/nat-n/poethepoet/blob/main/poethepoet/config.py#L201

Reply:
Hi Nat,

thanks a stack for all the information relevant to this topic. Your `iter_tasks` proposal looks good: Would it be more appropriate that it would actually yield `task` objects instead of names, and then there would be another `get_task_names` on top? Or would you prefer it to align to the same naming scheme, like `iter_tasknames`?

```python
from poethepoet.config import PoeConfig

def iter_tasks(project_root: str = "."):
    config = PoeConfig(cwd=project_root)
    config.load()
    for name, task in config.tasks.items():
        if name and name[0] != "_":
            yield name, task

def iter_tasknames(project_root: str = "."):
    for name, task in iter_tasks(project_root=project_root):
        yield name
```

With kind regards,
Andreas.

Reply:
Hey @amotl,

Hmm maybe, what would be the use of the task definition here? Task definitions are a little bit tricky to work with because the can have different types (string, list, dictionary) and it requires some contextual awareness to interpret them correctly. Either way I think one exposed helper function would be enough.

I favour simplicity, my main concern with this piece of code is that is remain about as lightweight at possible, because it's used for shell autocompletion.